src/passes/SimplifyGlobals.cpp

```diff
@@ -250,21 +320,79 @@ struct SimplifyGlobals : public Pass {
     }
   }
 
-  void removeWritesToUnreadGlobals() {
-    // Globals that are not exports and not read from can be eliminated
-    // (even if they are written to).
-    NameSet unreadGlobals;
+  // Removes writes from globals that will never do anything useful with the
+  // written value anyhow. Returns whether an addition iteration is necessary.
+  bool removeUnneededWrites() {
+    bool more = false;
+
+    // Globals that are not exports and not read from are unnecessary (even if
+    // they are written to). Likewise, globals that are only read from in order
+    // to write to themselves are unnecessary. First, find such globals.
+    NameSet unnecessaryGlobals;
     for (auto& global : module->globals) {
       auto& info = map[global->name];
-      if (!info.imported && !info.exported && !info.read) {
-        unreadGlobals.insert(global->name);
+
+      if (!info.written) {
+        // No writes occur here, so there is nothing for us to remove.
+        continue;
+      }
+
+      if (info.imported || info.exported) {
+        // If the global is observable from the outside, we can't do anythng
+        // here.
+        //
+        // TODO: optimize the case of an imported but immutable global, etc.
+        continue;
+      }
+
+      // We only ever read-only-to-write if all of our reads are done in places
+      // we identified as read-only-to-write. That is, we have eliminated the
+      // possibility of any other uses. (Technically, each read-to-write
+      // location might have more than one read since we did not count them, but
+      // only verified there was one read or more; but this is good enough as
+      // the common case has exactly one.)
```

Comment:
Not sure if I understand... Can you elaborate?

> each read-to-write location might have more than one read since we did not count them,

Didn't we count them? Aren't we checking the number of reads is the same as the number of `readOnlyToWrite` right below?
```cpp
bool onlyReadOnlyToWrite = (info.read == info.readOnlyToWrite);
```

Reply:
We might have something like this:
```js
if (global + global == 1) { global = 2 }
```
That has two reads of the global. If that is the entire program, then we'd count 2 reads and only 1 read-only-to-write. In that case we wouldn't optimize, although it happens to be safe.

We could count the number of gets inside each read-only-to-write, but I'm not sure it's worth the complexity. The j2cl pattern is pretty specific.